align/affineGap.go

```diff
@@ -1,259 +1,352 @@
+	route := make([]Cigar, 1)                                                     //initialie cigar route and routeIdx
+	var routeIdx_current int = 0
+	var k_inChecker_max, k_inChecker_min ColType //for affineGap, make a variable to hold k_inChecker_max, outside of for loop
+
+	for k1, k2 = int((score_highest_i-1)/checkersize_i), int((score_highest_j-1)/checkersize_j); k1 >= 0 && k2 >= 0; { //use a function of score_highest_i, score_highest_j, and checkersize to initialize the right k1 and k2, go to the correct checkerboard to start traceback
+
+		//Step 2: for a checkerboard, fill traceback, as well as find the max i and j when the checkerboard traceback ends. Since the trace matrix was initialized above, pass it back and forth to recycle the memory instead of creating another trace in Step 2 in every iteration
+		//for affineGap, step 2 is almost the same as constGap, just have 3 dimensions of k (0,1,2), use gapOpen and gapExtend instead of gapPen, add k_inChecker_max
+		trace, i_inChecker_max, j_inChecker_max, k_inChecker_max = fillTraceback_affineGap(alpha, beta, scores, gapOpen, gapExtend, checkersize_i, checkersize_j, score_highest_i, score_highest_j, trace_prep_i, trace_prep_j, k1, k2, i_inChecker_min, j_inChecker_min, trace)
+
+		//Step 3: for a checkerboard, use traceback to write cigar (update route and routeIdx), as well as find the min i and j when the checkerboard cigar ends
+		//for affineGap, step 3 is almost the same as constGap, just have 3 dimensions of k (0,1,2), add k_inChecker_min
+		route, routeIdx_current, i_inChecker_min, j_inChecker_min, k_inChecker_min = writeCigar_affineGap(trace, i_inChecker_max, j_inChecker_max, k_inChecker_max, route, routeIdx_current, i_inChecker_min, j_inChecker_min, k_inChecker_min)
+
+		//Use Step 3's i_inChecker_min and j_inChecker_min to find the next checkerboard to go to and where to start in that checkerboard (update k1, k2)
+		if i_inChecker_min < 0 && j_inChecker_min < 0 {
+			k1, k2 = k1-1, k2-1 //go to the next checkerboard with lower i, lower j
+		} else if i_inChecker_min < 0 { //but j_inChecker != 0
+			k1 -= 1 //go to the next checkerboard with lower i, same j
+		} else if j_inChecker_min < 0 { //but i_inChecker != 0
+			k2 -= 1 //go to the next checkerboard with lower j, same i
 		}
+
 	}
-	maxScore, route := affineTrace(mRowCurrent, mColumn, trace)
 
-	return maxScore, route
+	//Step 4: write the last cigar entry. Since this step doesn't need special features for affineGap, reuse constGap function
+	//This step is necessary because the row i=0 and the column j=0 are always stored in trace_prep, and never filled in in any checkerboard, but the cigar ends by the route going into 1 box/entry in either the row i=0 or the column j=0
+	if i_inChecker_min != -1 && j_inChecker_min == -1 { //indicating that Step 3 ended when k1 and k2 reached the smallest combination, and reached j=0, so the last cigar is a "2", e.g. if last cigar entry is the i=1, j=0 square
```

Comment:
Since the last square that was filled in the checkerboard reached j=0 (ran out of columns before running out of rows), and the beta sequence is on the x-axis (each column corresponds to a base in the beta sequence), reaching j=0 first means the last cigar should be a "deletion" in the beta sequence